### DISCUSSION TITLE
Choosing pins for spindle control

### DISCUSSION THREAD

**Discussion by fitch22** [grblHAL/core#59] [General]
While configuring a map file for the BTT SKR-2 board, I have defined pins for spindle enable, direction and pwm.  I chose pins designed for fans for enable and direction that can output 24V and both are PWM capable.  I chose a standard IO pin for the spindle pwm since I think a 5V signal is more typical here but I do have another fan output same as for enable and direction so I could use that instead.  I will also mention, that these fan outputs actually switch the GND side, not the + side, so if necessary it would be easy to use them with 5V or 12V just by feeding a different + side.  Also, these are easy connections to make to the board as they all have their own connectors.

What would be the best choice for the type of driving pin?  Is a pin capable of 24V appropriate, or would essentially everyone have to feed different voltage?

Also, for my 3018 CNC, the spindle speed is varied using pwm on the 24V fed to it.  Should I use SPINDLE_ENABLE_PIN or SPINDLE_PWM_PIN to feed the spindle?  I don't fully understand the use cases for enable vs pwm.

Thanks.

**Comment by kfmut** [comment]
I thought that stm32 driver has fixed pin for spindle pwm output(A8, IIRC). Has it changed now?

**Reply by fitch22** [reply]
It looks to me like there are lots of PWM pins, not just one.

**Reply by terjeio** [reply]
> What would be the best choice for the type of driving pin?

Push-pull is the best - depends on how the spindle input is designed. If open drain like the mostfets are a resistor might have to be added wired to the appropriate voltage, or a push-pull output chosen.

> Is a pin capable of 24V appropriate, or would essentially everyone have to feed different voltage?

It depends on the spindle. Some accepts the PWM signal directly, some needs a DC voltage - many of those 0-10V.

>  Should I use SPINDLE_ENABLE_PIN or SPINDLE_PWM_PIN to feed the spindle? I don't fully understand the use cases for enable vs pwm.

Again it depends on the spindle, some have two or three inputs for controlling them, some only one. the PWM output is used when the spindle speed can be controlled, spindle on possibly in combination with PWM and when speed control is not possible.

> I thought that stm32 driver has fixed pin for spindle pwm output(A8, IIRC).

There are four different STM32 drivers, F4xx currently has support code for A7 and A8. 

> Has it changed now?

> It looks to me like there are lots of PWM pins, not just one.

Support for using other pins can be added - but it requires code changes. The correct timer has to be selected and set up and the PWM output routed to the pin. There may be conflicts with other use of the timer in question making it harder and perhaps impossible to assign the desired pin. Adding support for selecting from all available pins takes a lot of time, time wasted if pins are never used for PWM. What I should have done is to add a compile time error if an usupported pin is selected...

**Reply by fitch22** [reply]
It looks like TIM1 is the only timer used for PWM.  Is that right?  And it looks like to keep things simple, only pins that can connect to TIM1 outputs can be used, is that right?

For the SKR-2, the most desirable pins to support PWM would be PD7, PB3 and PB4 as they all have high current outputs and can drive a 24V spindle directly.  PD7 does not appear to connect to any timer so that looks out.  PB3 connects to TIM2 and PB4 connects to TIM3.

First of all, could the general purpose timers like TIM2 and TIM3 even support PWM as it is done in grblHal?  And if they could, might one of these timers be changed to a different general purpose timer to free up one of TIM2 or TIM3?  Please know that I am not asking to do this, I just want to know if it even could be done and some idea of the work involved.

Thanks,
Bill

**Comment by kfmut** [comment]
@terjeio

> F4xx currently has support code for A7 and A8

Good to know that, thank you!

**Reply by terjeio** [reply]
Support for spindle PWM -> PB0 added to STM32F4xx driver. Will become available in the next commit. Other PWM capable pins routable to relevant Timer1 outputs will be fairly easy to add.

**Reply by terjeio** [reply]
> First of all, could the general purpose timers like TIM2 and TIM3 even support PWM as it is done in grblHal? And if they could, might one of these timers be changed to a different general purpose timer to free up one of TIM2 or TIM3? Please know that I am not asking to do this, I just want to know if it even could be done and some idea of the work involved.

IMO It is not neccesary to free up already allocated timers unless the pin selection dictates so, there are others to choose from depending on the F4 variant. If allocating a currently unused timer as an alternative for PWM then less code has to be changed and less testing is needed. If you have a good understanding of how to configure STM32 timers for PWM mode then it should not take too much time.
Be aware that is important that when stopping the PWM timer it should leave the output pin in a state that depends on whether the output is configured as inverted or not.

I have to come up with a better way to select the PWM pin, it is not possible(?) to use the preprocessor to check the validity of the port/pin combo as it is done now. I'll have to figure out a way, or do you have an idea of how it can be done?

**Reply by terjeio** [reply]
I have been sleeping on this and after checking the datasheets and STM code files I found by switching to defining the PWM port by its base address and some fancy token pasting it is possible to select predefined pins (in driver.h) for PWM output. TIM2 and TIM3 is currently used by PPI and spindle sync so can be used for PWM if not enabled. Here is part of the new code:

```
#ifdef SPINDLE_PWM_PORT_BASE

#if SPINDLE_PWM_PORT_BASE == GPIOA_BASE
  #if SPINDLE_PWM_PIN == 7 // PA7 - TIM1_CH1N
    #define SPINDLE_PWM_TIMER_N     1
    #define SPINDLE_PWM_TIMER_CH    1
    #define SPINDLE_PWM_TIMER_INV   1
    #define SPINDLE_PWM_TIMER_AF    1
  #elif SPINDLE_PWM_PIN == 8 // PA8 - TIM1_CH1
    #define SPINDLE_PWM_TIMER_N     1
    #define SPINDLE_PWM_TIMER_CH    1
    #define SPINDLE_PWM_TIMER_INV   0
    #define SPINDLE_PWM_TIMER_AF    1
  #endif
#elif SPINDLE_PWM_PORT_BASE == GPIOB_BASE
  #if SPINDLE_PWM_PIN == 0 // PB0 - TIM1_CH2N
    #define SPINDLE_PWM_TIMER_N     1
    #define SPINDLE_PWM_TIMER_CH    2
    #define SPINDLE_PWM_TIMER_INV   1
    #define SPINDLE_PWM_TIMER_AF    1
  #elif SPINDLE_PWM_PIN == 3 // PB3 - TIM2_CH2
    #define SPINDLE_PWM_TIMER_N     2
    #define SPINDLE_PWM_TIMER_CH    2
    #define SPINDLE_PWM_TIMER_INV   0
    #define SPINDLE_PWM_TIMER_AF    1
  #elif SPINDLE_PWM_PIN == 4 // PB4 - TIM2_CH2
    #define SPINDLE_PWM_TIMER_N     3
    #define SPINDLE_PWM_TIMER_CH    1
    #define SPINDLE_PWM_TIMER_INV   0
    #define SPINDLE_PWM_TIMER_AF    2
  #endif
#endif

#if SPINDLE_PWM_TIMER_CH == 1 || SPINDLE_PWM_TIMER_CH == 2
#define SPINDLE_PWM_CCR 1
#else
#define SPINDLE_PWM_CCR 2
#endif
#define SPINDLE_PWM_TIMER           timer(SPINDLE_PWM_TIMER_N)
#define SPINDLE_PWM_TIMER_CCR       timerCCR(SPINDLE_PWM_TIMER_N, SPINDLE_PWM_TIMER_CH)
#define SPINDLE_PWM_TIMER_CCMR      timerCCMR(SPINDLE_PWM_TIMER_N, SPINDLE_PWM_CCR)
#define SPINDLE_PWM_CCMR_OCM_SET    timerOCM(SPINDLE_PWM_CCR, SPINDLE_PWM_TIMER_CH)
#define SPINDLE_PWM_CCMR_OCM_CLR    timerOCMC(SPINDLE_PWM_CCR, SPINDLE_PWM_TIMER_CH)
#if SPINDLE_PWM_TIMER_INV
#define SPINDLE_PWM_CCER_EN         timerCCEN(SPINDLE_PWM_TIMER_CH, N)
#define SPINDLE_PWM_CCER_POL        timerCCP(SPINDLE_PWM_TIMER_CH, N)
#define SPINDLE_PWM_CR2_OIS         timerCR2OIS(SPINDLE_PWM_TIMER_CH, N)
#else
#define SPINDLE_PWM_CCER_EN         timerCCEN(SPINDLE_PWM_TIMER_CH, )
#define SPINDLE_PWM_CCER_POL        timerCCP(SPINDLE_PWM_TIMER_CH, )
#define SPINDLE_PWM_CR2_OIS         timerCR2OIS(SPINDLE_PWM_TIMER_CH, )
#endif

#define SPINDLE_PWM_PORT            ((GPIO_TypeDef *)SPINDLE_PWM_PORT_BASE)
#define SPINDLE_PWM_AF              timerAF(SPINDLE_PWM_TIMER_N, SPINDLE_PWM_TIMER_AF)

#endif // SPINDLE_PWM_PORT_BASE

#if defined(SPINDLE_PWM_PIN) && !defined(SPINDLE_PWM_TIMER_N)
#ifdef SPINDLE_PWM_PORT
#error Map spindle port by defining SPINDLE_PWM_PORT_BASE in the map file!
#else
#error Spindle PWM not supported on mapped pin!
#endif
#endif
```

The new code is tested with the STM4xx driver, I believe the new code can be used "as-is" or with little modification for other STM32 drivers as well. Note that adding a new PWM pin in driver.h does not require any changes to driver.c. - only testing is required.

**Reply by fitch22** [reply]
Terje,
I have tested my SKR-2 board using PB0 (my original SPINDLE_PWM_PIN) and PB3 and PB4 (desired pins that can control my spindle directly) with the latest code.  PB0 behaves as expected, I can vary the duty cycle from 0 to 100% and the frequency is 5KHz.

PB3 and PB4 work, the only issue is the frequency of the PWM is 2.5KHz instead of 5KHz.  I have not yet poked through the code to see where things are different.  You may know exactly where to look.

I have not tried to change the PWM frequency, but maybe I will try that.

Bill

**Reply by fitch22** [reply]
Timer 1 is on APHB2 and Timers 2 and 3 are on APHB1.  APHB1 (at least on my SKR-2) is clocked at half the frequency of APHB2.  This could explain the PWM frequency if this is not accounted for.

**Reply by terjeio** [reply]
> Timer 1 is on APHB2 and Timers 2 and 3 are on APHB1. APHB1 (at least on my SKR-2) is clocked at half the frequency of APHB2. This could explain the PWM frequency if this is not accounted for.

I does explain it - but there is no _quick_ fix as the supported MCUs differs in the clock tree configuration. I'll have to go through the datasheets to find out how to solve this.

**Reply by terjeio** [reply]
Try by replacing this code in [driver.c ](https://github.com/grblHAL/STM32F4xx/blob/master/Src/driver.c)(line 1271+):

```
            uint32_t prescaler = settings->spindle.pwm_freq > 4000.0f ? 1 : (settings->spindle.pwm_freq > 200.0f ? 12 : 25);

            spindle_precompute_pwm_values(&spindle_pwm, SystemCoreClock / prescaler);
```
with this:

```
            RCC_ClkInitTypeDef clock;
            uint32_t latency, prescaler = settings->spindle.pwm_freq > 4000.0f ? 1 : (settings->spindle.pwm_freq > 200.0f ? 12 : 25);

            HAL_RCC_GetClockConfig(&clock, &latency);

#if SPINDLE_PWM_TIMER_N == 1
            spindle_precompute_pwm_values(&spindle_pwm, (HAL_RCC_GetPCLK1Freq() * (clock.APB1CLKDivider == 0 ? 1 : 2)) / prescaler);
#else
            spindle_precompute_pwm_values(&spindle_pwm, (HAL_RCC_GetPCLK2Freq() * (clock.APB2CLKDivider == 0 ? 1 : 2)) / prescaler);
#endif
```

as a temporary workaround.

It would be better to determine the prescaler value, and possibly the clock divison value, from the actual timer clock frequency to cover a large range of PWM frequencies (50Hz - 20kHz). For later...

**Comment by fitch22** [comment]
Terje,
This is awesome.  I would be happy to test this and I think I can test with a variety of different pins.  Unfortunately, I am about to head off on family vacation and won’t be able to look at anything for 2 weeks.

Thanks!

**Comment by fitch22** [comment]
Terje,

Thanks for the quick update.  I though this was a pretty good fix, but I
cannot get it to work.  When using pwm on PB0 (uses Timer 1), what used to
be 5KHz is now 10KHz.  When using pwm on PB3 (uses Timer 2) I still get
2.5KHz.

 

I think I know what is going on, but cannot find where to fix it.

 

In the original code, the frequency passed to
spindle_precompute_pwm_values() was 168MHz.  For Timer 1, the period
calculated to be 33600 and with the clock frequency for timer 1 of 84MHz,
this generated a pwm frequency of 5KHz as it is supposed to.

 

With the new code and pin PB0 (uses Timer 1) the frequency passed to
spindle_precompute_pwm_values() is 84MHz and the period calculated is 16800.
Since Timer 1 still has the same clock, a period of half the size results in
double the frequency or 10KHz.  It seems to me that passing a frequency of
84MHz is correct, the previous code passing 168MHz is not correct.
Somewhere I cannot find is a scaling by 2 that fixed the old code and breaks
the new.

 

When I use PB3 (this uses Timer 2) both old and new code pass the same
frequency of 168MHz and as you would expect, it outputs the same pwm
frequency in both cases.  However, it is 2.5KHz and not the desired default
5KHz.

 

I will continue to debug this.  If I find the issue, I will post again.

 

Bill

**Reply by terjeio** [reply]
> I think I know what is going on, but cannot find where to fix it.

I finally found the correct clock routing in the RCC register documentation, Timer 1 runs off APB2, 2-5 off APB1 clock. So the code should be:

```
#if SPINDLE_PWM_TIMER_N == 1
            spindle_precompute_pwm_values(&spindle_pwm, (HAL_RCC_GetPCLK2Freq() * (clock.APB2CLKDivider == 0 ? 1 : 2)) / prescaler);
#else
            spindle_precompute_pwm_values(&spindle_pwm, (HAL_RCC_GetPCLK1Freq() * (clock.APB1CLKDivider == 0 ? 1 : 2)) / prescaler);
#endif
```

**Comment by fitch22** [comment]
Doh!  I should have seen this.  Thanks for your help, this is great.

Bill

**Comment by fitch22** [comment]
Terje,

Everything now works great.  Thank you again.

 

Here is a question for you.  One of my cnc machines is a little 3018.  It came with a Arduino based controller board that has connections for both a spindle and a laser.  On this board, the same pwm output from the MCU feeds both the spindle driver (24V) and the laser (5V).  Since my SKR-2 does not have similar circuitry, if one wanted to use a spindle, they would choose a pwm pin such as PB3 since it can switch 24V, and if one wanted to use a laser they would choose a pwm pin such as PB0 since it is just standard TTL.  However, it would be really nice for those that have both a spindle and a laser to be able to switch those pins without recompiling.  One idea is that when switching modes from spindle (Mode_Standard) to laser (Mode_Laser) the pwm pin could be switched.  The 2 pwm pins could be defined slightly differently in the map file, e.g. SPINDLE_PWM_PIN and LASER_PWM_PIN.

 

Would something like this be of general interest to the GRBL community?  What issues would this present to Lathe mode, or would Lathe mode want a LATHE_PWM_PIN as well?  I would be happy to make this change if there is any use for it besides me.

 

Bill

**Comment by terjeio** [comment]
> One idea is that when switching modes from spindle (Mode_Standard) to laser (Mode_Laser) the pwm pin could be switched.

I do that in the [huanyang spindle plugin](https://github.com/grblHAL/Plugins_spindle/blob/master/huanyang.c) by "hooking into" the `hal.settings_changed` function pointer and swapping spindle HAL pointers on a mode change. You can do the same in your board support code. 

> Would something like this be of general interest to the GRBL community?

Could be, but making it a generic feature will be harder - not all drivers/boards have the pins needed for a secondary PWM output. Some could just reroute the PWM signal with a relay or other signal switch controlled by a GPIO/ioport output?

> What issues would this present to Lathe mode, or would Lathe mode want a LATHE_PWM_PIN as well?

I do not think so, but can of course be done in a similar way as done by the huanyang code.

> I would be happy to make this change if there is any use for it besides me.

Add it in the board support code for now?
